### PR TITLE
Visualize keyless tables and views (for EF Core 3.0 only)

### LINKED ISSUE #279: Visualize Keyless tables and views when using EF Core 3.0

### PR DESCRIPTION
**Pull request by ErikEJ**
fixes #279

### CODE REVIEW

**Comment by ErikEJ** [comment]
@Herdo Would be very grateful if you could have a look at what I have done wrong here!

File: src/GUI/RevEng.Shared/TableInformationModel.cs [hunk 1]
```diff
@@ -47,20 +48,37 @@ public bool HasPrimaryKey
...
+                OnPropertyChanged();
+            }
+        }
+
         /// <summary>
         /// Initializes a new instance of the <see cref="TableInformationModel"/> class for a specific table.
         /// </summary>
         /// <param name="name">The table name.</param>
         /// <param name="hasPrimaryKey">Whether or not a primary key exists for the table.</param>
         /// <exception cref="ArgumentException"><paramref name="schema"/> or <paramref name="name"/> are null or only white spaces.</exception>
```
Comment (Herdo) [inline src/GUI/RevEng.Shared/TableInformationModel.cs]: Shoul add XML documentation here, what `isKeyless` is about in the constructor.

File: src/GUI/RevEng.Shared/TableInformationModel.cs [hunk 2]
```diff
@@ -47,20 +48,37 @@ public bool HasPrimaryKey
             }
         }
 
+        /// <summary>
+        /// Gets or sets whether the table or view is keyless - always false for EF Core 2.0.
+        /// </summary>
```
Comment (Herdo) [inline src/GUI/RevEng.Shared/TableInformationModel.cs]: Maybe add an `<example>` section below `<summary>` explaining the difference between `HasPrimaryKey` and `IsKeyless`.

File: src/GUI/RevEng.Shared/TableInformationModel.cs [hunk 3]
```diff
@@ -47,20 +48,37 @@ public bool HasPrimaryKey
...
         /// <param name="name">The table name.</param>
         /// <param name="hasPrimaryKey">Whether or not a primary key exists for the table.</param>
         /// <exception cref="ArgumentException"><paramref name="schema"/> or <paramref name="name"/> are null or only white spaces.</exception>
         public TableInformationModel(string name,
-                                     bool hasPrimaryKey)
+                                     bool hasPrimaryKey,
+                                     bool isKeyless = false)
         {
             if (string.IsNullOrWhiteSpace(name))
                 throw new ArgumentException(@"Value cannot be empty or only white spaces.", nameof(name));
```
Comment (Herdo) [inline src/GUI/RevEng.Shared/TableInformationModel.cs]: Is every combination of `HasPrimaryKey` and `IsKeyless` valid? If not, throw an exception in the constructor?

File: src/GUI/EFCorePowerTools/Helpers/RepositoryHelper.cs [hunk 4]
```diff
@@ -45,7 +45,7 @@ public static List<TableInformationModel> GetTablesFromRepository(DatabaseInfo d
                     var views = repository.GetAllViews();
                     foreach (var view in views)
                     {
-                        var info = new TableInformationModel(view.ViewName, true);
+                        var info = new TableInformationModel(view.ViewName, true, includeViews ? true : false);
```
Comment (Herdo) [inline src/GUI/EFCorePowerTools/Helpers/RepositoryHelper.cs]: Line 43 is `if (includeViews)` - therefore, you may just put `true` here.

File: src/GUI/EFCorePowerTools/Helpers/RepositoryHelper.cs [hunk 5]
```diff
@@ -36,7 +36,7 @@ public static List<TableInformationModel> GetTablesFromRepository(DatabaseInfo d
                 foreach (var table in tableList)
                 {
                     var hasPrimaryKey = allPks.Any(m => m.TableName == table);
-                    var info = new TableInformationModel(table, includeViews ? true : hasPrimaryKey);
+                    var info = new TableInformationModel(table, includeViews ? true : hasPrimaryKey, includeViews ? !hasPrimaryKey : false);
                     tables.Add(info);
                 }
```
Comment (Herdo) [inline src/GUI/EFCorePowerTools/Helpers/RepositoryHelper.cs]: It doesn't seem logical to evaluate `includeViews` for tables.

File: src/GUI/EFCorePowerTools/Helpers/EnvDTEHelper.cs [hunk 6]
```diff
@@ -245,7 +245,7 @@ internal static List<TableInformationModel> GetMysqlTableNames(string connection
                     foreach (string table in tables)
                     {
                         bool hasPrimaryKey = HasMysqlPrimaryKey(schema, table, mysqlConn);
-                        var info = new TableInformationModel(table, includeViews ? true : hasPrimaryKey);
+                        var info = new TableInformationModel(table, includeViews ? true : hasPrimaryKey, includeViews ? !hasPrimaryKey : false);
                         result.Add(info);
                     }
```
Comment (Herdo) [inline src/GUI/EFCorePowerTools/Helpers/EnvDTEHelper.cs]: It doesn't seem logical to evaluate `includeViews` for tables.

File: src/GUI/EFCorePowerTools/Helpers/EnvDTEHelper.cs [hunk 7]
```diff
@@ -216,7 +216,7 @@ internal static List<TableInformationModel> GetNpgsqlTableNames(string connectio
                     var viewsDataTable = npgsqlConn.GetSchema("Views");
                     foreach (DataRow row in viewsDataTable.Rows)
                     {
-                        var info = new TableInformationModel(row["table_schema"].ToString() + "." + row["table_name"].ToString(), true);
+                        var info = new TableInformationModel(row["table_schema"].ToString() + "." + row["table_name"].ToString(), true, includeViews ? true : false);
```
Comment (Herdo) [inline src/GUI/EFCorePowerTools/Helpers/EnvDTEHelper.cs]: Views have always a primary key, but may be keyless depending on `includeViews`?

File: src/GUI/EFCorePowerTools/Helpers/EnvDTEHelper.cs [hunk 8]
```diff
@@ -207,7 +207,7 @@ internal static List<TableInformationModel> GetNpgsqlTableNames(string connectio
                         && myRow.Field<string>("constraint_type") == "PRIMARY KEY")
                         .FirstOrDefault();
 
-                    var info = new TableInformationModel(row["table_schema"].ToString() + "." + row["table_name"].ToString(), includeViews ? true : primaryKey != null);
+                    var info = new TableInformationModel(row["table_schema"].ToString() + "." + row["table_name"].ToString(), includeViews ? true : primaryKey != null, includeViews ? primaryKey == null : false);
```
Comment (Herdo) [inline src/GUI/EFCorePowerTools/Helpers/EnvDTEHelper.cs]: It doesn't seem logical to evaluate `includeViews` for tables.

**Comment by ErikEJ** [comment]
Thanks a lot! a general comment about includeviews - it means "running in EF Core 3.0 mode" more than anything else.

**Comment by Herdo** [comment]
I'm picking one code snippet here:

The code from [src/GUI/EFCorePowerTools/Helpers/RepositoryHelper.cs](https://github.com/ErikEJ/EFCorePowerTools/pull/280/files/247ed18a8e44b5dcececd252be401d9c6235a241#diff-370726e109d3c328e69256c169504101), line 39 ...

```csharp
var info = new TableInformationModel(table,
                                     hasPrimaryKey: includeViews ? true : hasPrimaryKey,
                                     isKeyless: includeViews ? !hasPrimaryKey : false);
```
would be equal to this?

```cs
var info = new TableInformationModel(table,
                                     hasPrimaryKey: efcore30 ? true : hasPrimaryKey,
                                     isKeyless: efcore30 ? !hasPrimaryKey : false);
```

Therefore, when using EF Core 3.0...

- **tables always have primary keys?**
- tables are keyless when they don't have primary keys?

And when using EF Core 2.X...

- tables have primary keys when `hasPrimaryKey` is `true`?
- tables are never keyless?

This is confusing, to be honest.

**Comment by ErikEJ** [comment]
I think you are overthinking it. It is all about showing the " this is keyless" warning icon in the UI. Maybe just rename to: ShowKeylessWarningIcon ?

**Comment by Herdo** [comment]
> I think you are overthinking it. It is all about showing the " this is keyless" warning icon in the UI. Maybe just rename to: ShowKeylessWarningIcon ?

Maybe I am 😄 

To simplify it, I'd rather suggest to scrap the second parameter, and rename `HasPrimaryKey`? The result would be the same?

OFC, renaming the other to `ShowKeylessWarningIcon` could also work.

**Comment by ErikEJ** [comment]
That would not work, as it is used for other things in the UI

**Pull request merged**